### PR TITLE
WIP: Add backdrop modal widget

### PR DESCRIPTION
**Pull request by hSloan**

### CODE REVIEW

File: backend-db/Rhyolite/Backend/TaskWorker.hs [hunk 1]
```diff
@@ -1,4 +1,6 @@
 {-# LANGUAGE FlexibleContexts #-}
```
Comment (ali-abrar) [inline backend-db/Rhyolite/Backend/TaskWorker.hs]: Please separate the backdrop stuff from the unrelated changes (to tasks etc).

File: common/Rhyolite/App.hs [hunk 2]
```diff
@@ -34,6 +38,10 @@ singletonQuery k = QueryMorphism { _queryMorphism_mapQuery = MonoidalMap.singlet
                                  , _queryMorphism_mapQueryResult = MonoidalMap.findWithDefault mempty k
                                  }
 
+deriving instance FunctorMaybe (MonoidalMap k)
+instance Witherable (MonoidalMap k) where
+  wither f (MonoidalMap.MonoidalMap xs) = MonoidalMap.MonoidalMap <$> wither f xs
```
Comment (ali-abrar) [inline common/Rhyolite/App.hs]: This shouldn't be part of this PR

File: frontend/Rhyolite/Frontend/Widget.hs [hunk 3]
```diff
@@ -78,3 +82,25 @@ extensibleListWidgetWithSize n x0 xs0 addAtEnd itemWidget = do
           valuesMapD = joinDynThroughMap $ fmap (fmap snd) $ resultMapD
           valuesD = fmap Map.elems valuesMapD
   return valuesD
+
+-- | Widget used as a modal div for widgets that want to take some action when clicked anywhere but itself, such as dropdown widgets or the like.
+  -- The first argument is a CSS class name, the suggested CSS class styling for use of this widget is as follows:
+  -- position: fixed;
```
Comment (ali-abrar) [inline frontend/Rhyolite/Frontend/Widget.hs]: Can you provide a Clay value?

File: frontend/Rhyolite/Frontend/Widget.hs [hunk 4]
```diff
@@ -78,3 +82,25 @@ extensibleListWidgetWithSize n x0 xs0 addAtEnd itemWidget = do
...
+  -- top: 0;
+  -- bottom: 0;
+  -- right: 0;
+  -- left: 0;
+  -- z-index: 100;
```
Comment (3noch) [inline frontend/Rhyolite/Frontend/Widget.hs]: `z-index` only works for explicitly positioned elements. So the code here works but must be placed before all other DOM-building. As such it is far less convenient to use. Here's an example of the same idea, but where we introduce a monad to make it much more ergonomic (you can use `tellModal` from anywhere in the app and it will build the DOM at the bottom to give it natural stacking order on top of everything else).

https://gitlab.com/obsidian.systems/tezos-bake-monitor/blob/develop/tezos-bake-central/frontend/src/Frontend/Modal/Base.hs#L106-108

Also note in this example, we pass the close Event into the modal-building function so that it can choose to accept background clicks or not accept them as close events. It also accepts Esc key.

**Comment by ali-abrar** [comment]
I'm not sure that's right given the presence of `position: fixed`

File: frontend/Rhyolite/Frontend/Widget.hs (hunk 4 above)
Comment (ali-abrar) [reply frontend/Rhyolite/Frontend/Widget.hs]: I'm not sure that's true, given the presence of `position: fixed`

Comment (3noch) [reply frontend/Rhyolite/Frontend/Widget.hs]: Ah you're probably right for the commented example. I suppose the point is, if you don't pass the right CSS here it won't work as intended. There is a way to make it work regardless of the CSS however, and that's to use it at the bottom of the DOM. But the API is far too painful to use as-is when you do it that way, hence the example.

Comment (3noch) [reply frontend/Rhyolite/Frontend/Widget.hs]: At the very least I'd recommend we name the function less generally. It over promises with this name (more like `withBackdropWhenConfiguredCorrectly`). Even with explicitly positioned elements, I believe it will only work within that particular "stack" (the nuances escape me), so your backdrop still might not cover the whole page.

**Comment by ali-abrar** [comment]
@hSloan What's going on with this PR? It's been in WIP status for quite some time.

**Comment by 3noch** [comment]
I will upstream our modal code as a PR to provide an alternative to this.

**Comment by 3noch** [comment]
See #81

**Comment by 3noch** [comment]
#81 was merged.
